pkg/infradb/utils.go

```diff
@@ -0,0 +1,294 @@
+)
+
+func startReplayProcedure(componentName string) {
+	globalLock.Lock()
+
+	var deferErr error
+	var preSubscriber *actionbus.Subscriber
+
+	defer func() {
+		if deferErr != nil {
+			log.Println("startReplayProcedure(): The replay procedure has failed")
+			log.Println("startReplayProcedure(): unblocking the TaskManager to continue")
+			taskmanager.TaskMan.ReplayFinished()
+		}
+	}()
+
+	preSubscribers := actionbus.ABus.GetSubscribers("preReplay")
+
+	for _, preSub := range preSubscribers {
+		if preSub.Name == componentName {
+			preSubscriber = preSub
+			break
+		}
+	}
+
+	if preSubscriber == nil {
+		deferErr = fmt.Errorf("no pre-replay subscriber for %s", componentName)
+		log.Printf("startReplayProcedure(): Error %+v\n", deferErr)
+		return
+	}
+
+	// Notify the preReplay subscriber
+	actionData := actionbus.NewActionData()
+	deferErr = actionbus.ABus.Publish(actionData, preSubscriber)
+	if deferErr != nil {
+		log.Printf("startReplayProcedure(): Error %+v\n", deferErr)
+		return
+	}
+
+	// Waiting for the pre-replay procedure to finish
+	deferErr = <-actionData.ErrCh
+	close(actionData.ErrCh)
+
+	if deferErr != nil {
+		log.Printf("startReplayProcedure(): Error %+v\n", deferErr)
+		return
+	}
+
+	log.Printf("startReplayProcedure(): Component %s has successfully executed pre-replay steps", componentName)
+
+	objectTypesToReplay := getObjectTypesToReplay(componentName)
+
+	objectsToReplay, subsForReplay, err := gatherObjectsAndSubsToReplay(componentName, objectTypesToReplay)
+
+	if err != nil {
+		log.Printf("startReplayProcedure(): Error %+v\n", err)
+		deferErr = err
+		return
+	}
+
+	// Releasing the lock as all the operations in the DB has finished
+	globalLock.Unlock()
+
+	// Notify task manager to continue processing tasks as
+	// the replay of objects in the DB has finished
+	taskmanager.TaskMan.ReplayFinished()
+
+	createReplayTasks(objectsToReplay, subsForReplay)
+}
+
+// getObjectTypesToReplay collects all the types of object to be replayed
+// which are related to the component that called the replay.
+func getObjectTypesToReplay(componentName string) []string {
+	objectTypesToReplay := []string{}
+
+	bpSubs := eventbus.EBus.GetSubscribers("bridge-port")
+	sviSubs := eventbus.EBus.GetSubscribers("svi")
+	lbSubs := eventbus.EBus.GetSubscribers("logical-bridge")
+	vrfSubs := eventbus.EBus.GetSubscribers("vrf")
+
+	for _, vrfSub := range vrfSubs {
+		if vrfSub.Name == componentName {
+			objectTypesToReplay = append(objectTypesToReplay, "vrf")
+			break
+		}
+	}
+
+	for _, lbSub := range lbSubs {
+		if lbSub.Name == componentName {
+			objectTypesToReplay = append(objectTypesToReplay, "logical-bridge")
+			break
+		}
+	}
+
+	for _, sviSub := range sviSubs {
+		if sviSub.Name == componentName {
+			objectTypesToReplay = append(objectTypesToReplay, "svi")
+			break
+		}
+	}
+
+	for _, bpSub := range bpSubs {
+		if bpSub.Name == componentName {
+			objectTypesToReplay = append(objectTypesToReplay, "bridge-port")
```

Comment:
why bare strings are used for this purpose. Strings are dangerous since we can easily make a typo. It is much more reliable to use enums(int or string)

Reply:
I think you are right on this. We can include that in the whole restructuring effor into the future. I will open an issue to track this one because we have this problem I believe in several places into the code. I hope that is allright

Reply:
pls create an issue/add to the existing one